From ddc822d46853d67a59aea75df54ba5f3505088c5 Mon Sep 17 00:00:00 2001
From: Mingcong Bai <baimingcong@uniontech.com>
Date: Mon, 2 Sep 2024 10:19:00 +0800
Subject: [PATCH 099/103] AOSCOS: Revert "drm/amdgpu: fix contiguous handling
 for IB parsing v2"

This commit may have caused hardware decoding for H.264 to fail on GCN 1.0
and 2.0 cards (AMD Radeon 520, R9 390X, etc.). Whilst playing such videos,
mpv crashes with the following output:

  amdgpu: The CS has been rejected, see dmesg for more information (-22).

Kernel logs:

  [drm:amdgpu_uvd_pass2 [amdgpu]]: *ERROR* msg/fb buffer ff019e0000-
  ff019e4000 out of 256MB segment!

Issue currently tracked at:

  https://gitlab.freedesktop.org/drm/amd/-/issues/3599

This reverts commit f3572db3c049b4d32bb5ba77ad5305616c44c7c1.

Link: https://gitlab.freedesktop.org/drm/amd/-/issues/3599#note_2550915
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 drivers/gpu/drm/amd/amdgpu/amdgpu_cs.c | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_cs.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_cs.c
index 6dfdff58bffd..56966d4f5bb2 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_cs.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_cs.c
@@ -1781,7 +1781,7 @@ int amdgpu_cs_find_mapping(struct amdgpu_cs_parser *parser,
 	struct ttm_operation_ctx ctx = { false, false };
 	struct amdgpu_vm *vm = &fpriv->vm;
 	struct amdgpu_bo_va_mapping *mapping;
-	int i, r;
+	int r;
 
 	addr /= AMDGPU_GPU_PAGE_SIZE;
 
@@ -1796,13 +1796,13 @@ int amdgpu_cs_find_mapping(struct amdgpu_cs_parser *parser,
 	if (dma_resv_locking_ctx((*bo)->tbo.base.resv) != &parser->exec.ticket)
 		return -EINVAL;
 
-	(*bo)->flags |= AMDGPU_GEM_CREATE_VRAM_CONTIGUOUS;
-	amdgpu_bo_placement_from_domain(*bo, (*bo)->allowed_domains);
-	for (i = 0; i < (*bo)->placement.num_placement; i++)
-		(*bo)->placements[i].flags |= TTM_PL_FLAG_CONTIGUOUS;
-	r = ttm_bo_validate(&(*bo)->tbo, &(*bo)->placement, &ctx);
-	if (r)
-		return r;
+	if (!((*bo)->flags & AMDGPU_GEM_CREATE_VRAM_CONTIGUOUS)) {
+		(*bo)->flags |= AMDGPU_GEM_CREATE_VRAM_CONTIGUOUS;
+		amdgpu_bo_placement_from_domain(*bo, (*bo)->allowed_domains);
+		r = ttm_bo_validate(&(*bo)->tbo, &(*bo)->placement, &ctx);
+		if (r)
+			return r;
+	}
 
 	return amdgpu_ttm_alloc_gart(&(*bo)->tbo);
 }
-- 
2.47.0.rc0

